pkg/cmd/roachtest/test_runner.go

```diff
@@ -556,114 +556,72 @@ func (r *testRunner) runWorker(
 			}
+			// We use a context that can't be canceled for the Destroy().
+			c.Destroy(context.Background(), closeLogger, l)
+			wStatus.SetCluster(nil)
+			c = nil
 		}
-		arch := testToRun.spec.Cluster.Arch
-		// N.B. local cluster can mix different CPU architectures via emulation; e.g., mac silicon running x86.
-		if testToRun.canReuseCluster && c != nil && c.spec.Cloud != spec.Local {
-			// We're reusing a non-local cluster, so we must use the same arch.
-			arch = c.arch
+
+		if testToRun.noWork {
+			shout(ctx, l, stdout, "no work remaining; runWorker is bailing out...")
+			return nil
 		}
-		if arch == "" {
-			// CPU architecture is unspecified, choose one according to the probability distribution.
-			arch = vm.ArchAMD64
-			if prng.Float64() < arm64Probability {
-				arch = vm.ArchARM64
-			} else if prng.Float64() < fipsProbability {
-				// N.B. branch is taken with probability (1 - arm64Probability) * fipsProbability which is P(fips | amd64).
-				// N.B. FIPS is only supported on 'amd64' at this time.
-				arch = vm.ArchFIPS
-			}
-			if testToRun.spec.Benchmark && testToRun.spec.Cluster.Cloud != spec.Local {
-				// TODO(srosenberg): enable after https://github.com/cockroachdb/cockroach/issues/104213
-				l.PrintfCtx(ctx, "Disabling randomly chosen arch=%q, %s", arch, testToRun.spec.Name)
-				arch = vm.ArchAMD64
-			}
-			l.PrintfCtx(ctx, "Using randomly chosen arch=%q, %s", arch, testToRun.spec.Name)
+
+		// From this point onward, c != nil iff we are reusing the cluster.
+
+		var arch vm.CPUArch
+		if c != nil && !c.IsLocal() {
+			// We are reusing a non-local cluster. We have already determined that its
+			// architecture is acceptable for the test (from the fact that the
+			// previous cluster spec had the same arch).
+			//
+			// Note that we treat local clusters differently because (in the case of
+			// Apple M1/M2) it can run multiple architectures.
+			// TODO(radu): this is not true of Intel and/or linux hosts, we should
+			// somehow determine the capabilities at runtime.
+			arch = c.arch
 		} else {
-			l.PrintfCtx(ctx, "Using specified arch=%q, %s", arch, testToRun.spec.Name)
-		}
-		// N.B. if canReuseCluster is false, then the previous cluster has been destroyed; new one will be created below.
-		if testToRun.canReuseCluster && c != nil && c.arch != arch {
-			// Non-local cluster that's being reused must have the same architecture as was ensured above.
-			if c.spec.Cloud != spec.Local {
-				return errors.Newf("infeasible path: non-local cluster arch=%q differs from selected arch=%q", c.arch, arch)
+			arch = archForTest(ctx, l, testToRun.spec)
+			if c != nil {
+				// Switch architecture of local cluster (see above).
+				c.arch = arch
 			}
-			// Local cluster is now reused to emulate a different CPU architecture.
-			c.arch = arch
 		}
 
 		// Verify that required native libraries are available.
+		//
+		// TODO(radu): the arch is not guaranteed and another arch can be selected
```

Comment:
Good point! This verification step should probably be moved into cluster creation as a (user-specified) precondition.